scripts/tundra/tools/msvc-vscommon-next.lua

```diff
@@ -0,0 +1,204 @@
+  local sdk_version = options.SdkVersion --default from vcvarsall.bat
+  local installation_path = options.InstallationPath or "C:\\Program Files (x86)\\Microsoft Visual Studio"
+  local vs_product = options.Product
+  
+  if arch_tuple == nil then
+    error("unsupported host/target architecture " .. arch_tuple)
+  end
+
+  -- don't assume that everyone is using the same edition of Visual Studio
+  -- unless a specific edition (i.e. product) has been requested, look for
+  -- the first setup that has Visual C++ tooling installed
+
+  local vcvarsall_bat
+
+  if vs_product == nil then
+    for _, vs_product in pairs(vs_products) do
+      vcvarsall_bat = find_vc_tools(installation_path, version, vs_product)
+      if vcvarsall_bat ~= nil then
+        break
+      end
+    end
+  else
+    vcvarsall_bat = find_vc_tools(installation_path, version, vs_product)
+  end
+
+  if vcvarsall_bat == nil then
+    error("cannot find installation of Visual Studio")
+  end
+
+  -- now to the fun part, the vcvarsall.bat script is quite customizable
+  -- we can request the environment from it, let it do the hard part and
+  -- then diff and merge things back into our environment
+
+  -- Syntax:
+  --     vcvarsall.bat [arch] [platform_type] [winsdk_version] [-vcvars_ver=vc_version]
+  -- where :
+  --     [arch]: x86 | amd64 | x86_amd64 | x86_arm | x86_arm64 | amd64_x86 | amd64_arm | amd64_arm64
+  --     [platform_type]: {empty} | store | uwp
+  --     [winsdk_version] : full Windows 10 SDK number (e.g. 10.0.10240.0) or "8.1" to use the Windows 8.1 SDK.
+  --     [vc_version] : "14.0" for VC++ 2015 Compiler Toolset | {empty} for default VS 2017 VC++ compiler toolset
+  -- 
+  -- The store parameter sets environment variables to support Universal Windows Platform application
+  -- development and is an alias for 'uwp'.
+  -- 
+  -- For example:
+  --     vcvarsall.bat x86_amd64
+  --     vcvarsall.bat x86_amd64 10.0.10240.0
+  --     vcvarsall.bat x86_arm uwp 10.0.10240.0
+  --     vcvarsall.bat x86_arm onecore 10.0.10240.0 -vcvars_ver=14.0
+  --     vcvarsall.bat x64 8.1
+  --     vcvarsall.bat x64 store 8.1
+
+  local vcvars_args = { arch_tuple }
+
+  if platform_type ~= nil then
+    vcvars_args[#vcvars_args+1] = platform_type
+  end
+
+  if sdk_version ~= nil then
+    vcvars_args[#vcvars_args+1] = sdk_version
+  end
+
+  local vcvars = {}
+
+  local command = "\"" .. vcvarsall_bat .. "\" " .. table.concat(vcvars_args, " ") .. " 1>NUL && set"
+  print(command)
+  local vcvars_p = io.popen(command)
+  while true do
+    local ln = vcvars_p:read()
+    if ln == nil then
+      break
+    end
+    local split_at = ln:find("=")
+    local k = ln:sub(1, split_at-1):upper()
+    local v = ln:sub(split_at+1)
+    vcvars[k] = v
+  end
+  vcvars_p:close()
+
+  --print(table.concat(keys(vcvars), " "))
+
+  -- Force MSPDBSRV.EXE (fixes issues with cl.exe running in parallel and corrupting PDB files)
+  env:set("CCOPTS", "/FS")
+  env:set("CXXOPTS", "/FS")
+
```

Comment:
The simplest and certainly crude way I could think of to merge the environment but I can make it a bit more surgical. For example, we don't have to copy the entire PATH variable as is.